### ISSUE TITLE
VTab: filter_pushdown support

### ISSUE THREAD

**Issue by joewalnes** [duckdb/duckdb-rs#401]
Request to include filter pushdown support for vtab extensions in Rust bindings.

This would allow things like the postgres extension (C++) to be built in Rust - allowing WHERE clause parameters to be propagated. I have a similar need, and would love to use Rust for this.

I'm happy to contribute this, if project is happy with the idea.

### Enabling

```rust
impl TableFunction {
    ...
    pub fn supports_filter_pushdown(&self, supports: bool) -> &Self { ... }
    ...
}

pub trait VTab: Sized {
    ...
    fn supports_filter_pushdown() -> bool { false }
    ...
}
```

(Later: consider renaming the existing `supports_pushdown()` to `supports_projection_pushdown()`. In the [C++ API](https://github.com/duckdb/duckdb/blob/fb7701fec0caf058f295aea6f5b8ef91b1d9a942/src/include/duckdb/function/table_function.hpp#L353-L358) it's named more explicitly to avoid confusion.)

### Accessing

```rust
impl InitInfo {
    ...
    pub fn get_filters(&self) -> [(idx_t, TableFilter)] { ... }
    ...
}
```

### Data model

This is roughly equivalent to C++ API, but far more concise with Rust's algebraic data-types 

```rust
#[derive(Debug)]
pub enum TableFilter {
    Constant {
        expression_type: ExpressionType,
        value: Value,
    },
    IsNull,
    IsNotNull,
    Or(Vec<TableFilter>),
    And(Vec<TableFilter>),
    StructExtract {
        child_name: String,
        child_filter: Box<TableFilter>,
    },
    Optional(Box<TableFilter>),
}

impl TableFilter {
    pub fn evaluate(&self, value: &Value) -> bool { ... }
}

#[derive(Debug)]
pub enum ExpressionType {
    Equal,
    NotEqual,
    LessThan,
    GreaterThan,
    LessThanOrEqual,
    GreaterThanOrEqual,
}
```

Note: There are many many more ExpressionTypes in the [C++ API,](https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/enums/expression_type.hpp) but those 6 are the most practical. eg. see [postgres_scanner](https://github.com/duckdb/postgres_scanner/blob/03eaed75f0ec5500609b7a97aa05468493b229d1/src/postgres_filter_pushdown.cpp#L18-L33) usage.

**Comment by mlafeldt** [comment]
Thanks for bringing this up - it's a legitimate gap.

Unfortunately, DuckDB's C API doesn't currently support filter pushdown, so implementation on the Rust side won't be possible until we've extended the API accordingly.

**Comment by joewalnes** [comment]
I filed this back in 2024, parked the work, but recently found a need to come back to it, so keen to help!

It seems there's been multiple attempt to add filter pushdown to the C API, but these were never merged:
https://github.com/duckdb/duckdb/pull/14591
https://github.com/duckdb/duckdb/pull/19093

@taniabogatsch reasoning is while it matches the current C++ API, there's a proposed new C++ API that it should match instead. (Not sure of status of this - any update?)
https://github.com/duckdb/duckdb/pull/19093#issuecomment-3355495251

**Comment by taniabogatsch** [comment]
Hi @joewalnes - @kryonix has spent a lot of effort removing explicit table filters from DuckDB's core code, instead using expressions (see his PRs in duckdb/duckdb). We're also currently working on shipping a C API V2 with DuckDB's upcoming `v2` release, which will have filter pushdown support. We've announced this on DuckCon, and will soon-ish move the work we've done so far into duckdb/duckdb (once we have feature parity with the existing C API).

**Comment by taniabogatsch** [comment]
Once it lands in duckdb/duckdb, we'd be happy to get your take on the V2 API, and your thoughts on whether it works for your use case.
